utilities/xmlrpcpp/src/XmlRpcDispatch.cpp

```diff
@@ -11,9 +11,57 @@
 
 #if defined(_WINDOWS)
 # include <winsock2.h>
-static inline int poll( struct pollfd *pfd, int nfds, int timeout)
+static int poll( struct pollfd *pfd, int nfds, int timeout)
 {
-  return WSAPoll(pfd, nfds, timeout);
+  // workaround: "Windows 8 Bugs 309411 – WSAPoll does not report failed connections"
+  // https://curl.haxx.se/mail/lib-2012-10/0038.html
+  // the following logic is to use select() to check all writable socket connnection status.
+  // if all the sockets to be checked are not connected, it reports SOCKET_ERROR to
+  // error out, instead of going to WSAPoll() which causes infinitely wait situation.
+  FD_SET writable;
+  FD_SET error;
+  FD_ZERO(&writable);
+  FD_ZERO(&error);
+  for (int i = 0; i < nfds; ++i)
+  {
+    if (pfd[i].events & POLLOUT)
+    {
+      FD_SET(pfd[i].fd, &writable);
+      FD_SET(pfd[i].fd, &error);
+    }
+  }
+
+  int connectionError = 0;
+  if (writable.fd_count > 0)
+  {
+    int result = select(0, nullptr, &writable, &error, nullptr);
+    if (SOCKET_ERROR == result)
+    {
+      return SOCKET_ERROR;
+    }
+
+    if (0 != result)
+    {
+      for (int i = 0; i < nfds; ++i)
+      {
+        if ((pfd[i].events & POLLOUT) &&
+            (FD_ISSET(pfd[i].fd, &error)))
+        {
+          connectionError++;
+        }
+      }
+    }
+  }
+
+  if (connectionError == nfds)
+  {
+    // error out if all sockets are failed to connect.
+    return SOCKET_ERROR;
+  }
+  else
+  {
+    return WSAPoll(pfd, nfds, timeout);
+  }
 }
 
 # define USE_FTIME
```

Comment:
I know it's not directly related here, but in the process of reviewing this, I noticed some if/endif mismatching going on. The `#endif` at the bottom of the file is labeled for `USE_FTIME` but I believe actually corresponds to `WINDOWS`.

@seanyen Do you mind looking into this and correcting the comments if applicable.

Reply:
By checking the [commit history](https://github.com/ros/ros_comm/commit/79bfbe8f05fcf73840c88b14b6f80ee48c5a78b9), it seems to me that `USE_FTIME` was first introduced because `Windows` doesn't have `gettimeofday` so it ended up with diverged implementations. However, two years ago, a [PR](https://github.com/ros/ros_comm/pull/1249) was merged and now it is in favor of `ros::ros_steadytime`.

I would say perhaps we should see if there is any opportunity to converge the implementations now, since `ros::ros_steadytime` is a portable thing and theoretically it should behave across the platforms.

What do you think?

Reply:
Since this is orthogonal please follow up on this separately form this PR.